Normalize mobile usernames and ensure does not exist when creating via API

gherceg opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
https://dimagi-dev.atlassian.net/browse/SAAS-12586

This is the short term solution. The ideal long term solution is to unify mobile user creation in the UI and the API so that validation happens in one place. This has been sitting on my laptop locally, so I figured if I don't have time to get to the long term solution just yet, might as well PR this. Created [this ticket](https://dimagi-dev.atlassian.net/browse/SAAS-13587) to track the long term solution.
## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->
Locally tested and automated test coverage. Will test again on staging.
### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
Added automated tests in this PR to verify changes.
### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->


### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

gherceg committed:
normalize username before creating user

gherceg committed:
raise exception if normalized username already exists

millerdev commented on corehq/apps/api/resources/v0_5.py:
```diff
@@ -222,10 +222,18 @@ def get_resource_uri(self, bundle_or_obj=None, url_name='api_dispatch_detail'):
+        try:
+            normalized_username = normalize_username(bundle.data['username'], domain=kwargs['domain'])
+        except ValidationError as e:
+            raise BadRequest(str(e))
+
+        if user_exists(normalized_username).exists:
+            raise BadRequest(f'Username {normalized_username} already exists.')
```
Should these error messages be translated?

gherceg replied:
That's a good point. I translated the error message if a user already exists in 7fee083ba2f5e9d36b5f69ee530407d0cb7f32a5, but held off the validation error message. The ValidationError could be raised from different spots, and from what I can see are not translated. I don't want to add a generic translated error message here, because I think a more specific untranslated error message is more helpful than a translated generic error message. Definitely something that could be improved though! Let me know if you think this should be a blocker.

millerdev replied:
I think it's fine to merge this as is since it's an improvement. Fixing translations can be done in a follow-up PR.

millerdev reviewed (approved):
Looks good. Not sure if this should block on translations.

gherceg committed:
translate duplicate username error message

snopoke reviewed (approved):
I'm going to approve this PR since it is an improvement but I also notice that the user create view has a lot of additional checks. It would be nice to reuse that code if possible: https://github.com/dimagi/commcare-hq/blob/d1050530352abeeb0a662b9f887753dc410e911a/corehq/apps/users/views/mobile/users.py#L706-L740

gherceg committed:
consolidate mobile username checks

gherceg commented:
Apologies for the force push, but it was only related to the most recent commit, 7aba15b77a8e76c8babef641904b675554de946b which had not yet been reviewed. Everything else was untouched!

gherceg committed:
fix failing test

gherceg committed:
handle none username or domain

gherceg committed:
removed unused import

gherceg commented:
In 0d68d06fcc10af71de07c2c5e1fc5deb7c7a474c, I addressed `username` or `domain` being None. `normalize_username` handled domain being None, but I couldn't think of when this would be valid. The only use of `normalize_username` where domain was none that I saw was in `test_choice_provider.py`, so it seems good to remove that case entirely.

snopoke commented on corehq/apps/users/util.py:
```diff
@@ -45,6 +51,40 @@
+    except InvalidUsernameException:
+        error = _("Username may not contain special characters.")
+        if '..' in username:
+            error = _("Username may not contain consecutive . (period).")
```
```suggestion
            error = _("Username may not contain consecutive '.' (period).")
```

snopoke commented on corehq/apps/users/util.py:
```diff
@@ -45,6 +51,40 @@
+        if '..' in username:
+            error = _("Username may not contain consecutive . (period).")
+        elif username.endswith('.'):
+            error = _("Username may not end with a . (period).")
```
```suggestion
            error = _("Username may not end with a '.' (period).")
```

snopoke commented on corehq/apps/users/validation.py:
```diff
@@ -0,0 +1,47 @@
+
+
+def validate_mobile_username(username, domain):
+    if username is None:
+        raise InvalidUsernameException
+    if domain is None:
+        raise InvalidDomainException
```
what about just a 'ValueError' here instead of custom exceptions?

snopoke commented on corehq/apps/users/validation.py:
```diff
@@ -0,0 +1,47 @@
+def _check_for_reserved_usernames(username):
+    reserved_usernames = ['admin', 'demo_user']
+    if username in reserved_usernames:
+        raise ReservedUsernameException
```
It's good practice to pass the bad value into the exception in case it is needed later. For example if there was code that was validating a list of usernames.
```suggestion
        raise ReservedUsernameException(username)
```

gherceg replied:
See See e9bb80ded2dd47c3a2f1c608fe2545524ed91b59

snopoke commented on corehq/apps/users/validation.py:
```diff
@@ -0,0 +1,47 @@
+    try:
+        validate_email(username)
+    except ValidationError:
+        raise InvalidUsernameException
```
```suggestion
        raise InvalidUsernameException(username)
```

gherceg replied:
See e9bb80ded2dd47c3a2f1c608fe2545524ed91b59

snopoke commented on corehq/apps/users/validation.py:
```diff
@@ -0,0 +1,47 @@
+        raise InvalidUsernameException
+
+    if BAD_MOBILE_USERNAME_REGEX.search(username) is not None:
+        raise InvalidUsernameException
```
```suggestion
        raise InvalidUsernameException(username)
```

gherceg replied:
See e9bb80ded2dd47c3a2f1c608fe2545524ed91b59

snopoke commented on corehq/apps/users/validation.py:
```diff
@@ -0,0 +1,47 @@
+        raise InvalidDomainException
+
+    _check_for_reserved_usernames(username)
+    username_as_email = format_username(username, domain)
+    _ensure_valid_username(username_as_email)
```
I see the order here has changed. Is it possible that the 'username_as_email' won't be a valid email address? For example if the domain name is somehow invalid as an actual domain?

Previously we were using 'example.com' to validate the username, just wondering if there was a reason for doing it that way.

gherceg replied:
Good point. Interestingly enough, in `normalize_username` we format the username with the domain first, and then call the same `validate_email` method:
https://github.com/dimagi/commcare-hq/blob/2968eeac63baed5dc1f088c907d5d00cdb8db06c/corehq/apps/users/util.py#L71-L73

That was added in 2011, whereas the "@example.com" version in `check_username` was added in 2018, so that portion of this check seems safe. 

However the `BAD_MOBILE_USERNAME_REGEX` is only used in check_usernames. I tested this locally by creating a domain "my%domain%", which is saved as "my-domain" to be used in URLs, and it looks like that url version is what is used when accessing a domain's name. I think this is safe to keep as is, but let me know if you have any concerns.

snopoke replied:
Thanks, that makes sense. Re the domain name - I agree that should be safe. I think the user input value is saved in `domain.hr_name` (human readable name).

snopoke commented on corehq/apps/users/tests/test_username_validation.py:
```diff
@@ -0,0 +1,135 @@
+        cls.domain_obj = create_domain('test-domain')
+        cls.addClassCleanup(cls.domain_obj.delete)
+
+    def setUp(self) -> None:
+        super().setUp()
```
nit: type hint and super call are unnecessary
```suggestion
    def setUp(self):
```

gherceg replied:
Ended up removing the use of `setUp` entirely so no longer relevant, but will keep in mind for the future. Thanks.

snopoke commented on corehq/apps/users/tests/test_username_validation.py:
```diff
@@ -0,0 +1,135 @@
+
+    def setUp(self) -> None:
+        super().setUp()
+        self.user = CommCareUser.create(self.domain, 'test-user@test-domain.commcarehq.org', 'abc123', None, None)
```
does this need to be in `setUp`? It will get deleted and re-created for each test case. I don't see any need for that so I think it could be moved to the `setUpClass`

gherceg replied:
See e9bb80ded2dd47c3a2f1c608fe2545524ed91b59

snopoke commented on corehq/apps/users/tests/test_username_validation.py:
```diff
@@ -0,0 +1,135 @@
+        cls.domain_obj = create_domain('test-domain')
+        cls.addClassCleanup(cls.domain_obj.delete)
+
+    def setUp(self) -> None:
+        super().setUp()
+        self.user = CommCareUser.create(self.domain, 'test-user@test-domain.commcarehq.org', 'abc123', None, None)
+        self.addCleanup(self.user.delete, self.domain, None)
```
same comment as above

snopoke commented on corehq/apps/users/tests/test_util.py:
```diff
@@ -187,3 +190,84 @@ def test_user_is_not_deactivated_and_no_logs(self):
+        cls.domain_obj = create_domain('test-domain')
+        cls.addClassCleanup(cls.domain_obj.delete)
+
+    def setUp(self) -> None:
+        super().setUp()
+        self.user = CommCareUser.create(self.domain, 'test-user@test-domain.commcarehq.org', 'abc123', None, None)
+        self.addCleanup(self.user.delete, self.domain, None)
```
same comment as above

snopoke commented on corehq/apps/users/tests/test_util.py:
```diff
@@ -187,3 +190,84 @@ def test_user_is_not_deactivated_and_no_logs(self):
+        self.assertEqual(username, 'test-user-1@test-domain.commcarehq.org')
+
+    def test_invalid_username_double_period_message(self):
+        try:
+            generate_mobile_username('test..user', self.domain)
+        except ValidationError as e:
+            self.assertEqual(e.message, 'Username may not contain consecutive . (period).')
+        else:
+            self.fail(f'Expected raised exception: {ValidationError}')
```
nit: The usual way to validate exception handling is as follows:
```suggestion
        with self.assertRaises(ValidationError) as cm:
            generate_mobile_username('test..user', self.domain)
        self.assertEqual(cm.exception.message, 'Username may not contain consecutive . (period).')
```

gherceg replied:
See e9bb80ded2dd47c3a2f1c608fe2545524ed91b59

snopoke commented on corehq/apps/users/util.py:
```diff
@@ -45,6 +51,40 @@
+    except ReservedUsernameException:
+        error = _("Username '{}' is reserved.").format(username)
+    except InvalidDomainException:
+        error = _("Domain is required.")
```
I don't think this should be handled. The 'domain' argument comes from the URL path so it should never be None. I would suggest not handling this error here at all and letting it cause a 500. If it ever is None then this message won't help the user resolve the issue.

gherceg replied:
That sounds good to me. I struggled to think of a scenario where the domain would be None.

snopoke reviewed (commented):
Thanks for the refactor and tests. Mostly looks good, just a few small comments.

gherceg committed:
address PR feedback

- cleanup error messages
- create user in class setup for tests
- raise ValueError instead of custom exception where it makes sense
- pass username into exceptions
- cleanup exception message validation logic

snopoke commented on corehq/apps/users/util.py:
```diff
@@ -45,6 +50,38 @@
+            error = _("Username '{}' is already taken.").format(username)
+    except ReservedUsernameException:
+        error = _("Username '{}' is reserved.").format(username)
+    finally:
```
with the change to `ValueError` it looks like a blank username will raise a 500. I think you need to handle the `ValueError` here as well.

gherceg replied:
See 61d8980534b74d4d3380c060f0a11008ff6c30e6

snopoke commented on corehq/apps/users/validation.py:
```diff
@@ -0,0 +1,52 @@
+    :return: str, email formatted mobile username
+    Example: validate_mobile_username('username', 'domain') -> 'username@domain.commcarehq.org'
+    """
+    if username is None:
```
should this also check if the username is blank?

gherceg replied:
That gets caught by the validation and is tested in `test_empty_username_raises_exception` which throws an `InvalidUsernameException`. If there are better reasons to put that check here, I'm open to it.

snopoke replied:
OK that's fine. Just wanted to make sure it was covered.

snopoke reviewed (commented):
Thanks for the cleanup, couple of followup comments.

gherceg committed:
handle ValueError

snopoke commented on corehq/apps/users/util.py:
```diff
@@ -45,6 +50,40 @@
+    except ReservedUsernameException:
+        error = _("Username '{}' is reserved.").format(username)
+    except ValueError:
+        error = _("Must specify a username.")
```
nit: consistent language in errors
```suggestion
        error = _("Username is required")
```

gherceg committed:
update error message if username is none

millerdev commented on corehq/apps/users/validation.py:
```diff
@@ -0,0 +1,52 @@
+    Example: validate_mobile_username('username', 'domain') -> 'username@domain.commcarehq.org'
+    """
+    if username is None:
+        raise ValueError
```
This is a generic error that could be raised in other places for other reasons. What do you think of using a more specific error type to indicate that a username is required?

snopoke replied:
This was my idea to use value error but in hindsight probably better to use a custom exception. Sorry @gherceg

gherceg replied:
No worries! I just reused the InvalidUsernameException.

millerdev reviewed (commented):
It does seem like a good idea to avoid double formatting if the username is already "email formatted." Do you know if people have been (and may be currently) submitting email-formatted usernames to the API?

gherceg committed:
use invalid username exception for empty username

gherceg commented:
> Do you know if people have been (and may be currently) submitting email-formatted usernames to the API?

This is a good question that I don't know the answer to. The UI very clearly does not allow email-formatted usernames as the InvalidUsernameException is thrown due to the use of special characters, and based on that I think it would be best to be consistent with the API as well. However, it would still be prudent to verify how the mobile username api has been used recently to see if this would disrupt any workflows, so I'll look into that.

gherceg commented:
I was unable to obtain logs on how username's are typically formatted for create requests since they live within the body of a POST. I reached out to CS to see if they had any insights but they don't train users on the API since it is usually devs who interact with the API. I noticed the [API docs](https://confluence.dimagi.com/pages/viewpage.action?pageId=22708569) do imply either format is accepted, so I'm assuming that has been the expectation up until this point.

At this point, it seems more important to get _something_ out because the fact that we have no validation around the username is concerning. I am still in favor of consistency between the API and the UI, and also reducing complexity for the user of the API. I can't think of a good reason to support the email formatted username, if the first part is enough.

raytula commented:
Note that new user creation is currently broken in our production deployment of CommCare in Guatemala, because API requests are returning an HTTP 400 error when we POST something that includes the domain name as the user name.   I'll adjust the code on our end, as I recently needed to do for updating users; however, I wanted to let you know that these unexpected changes in behaviour can have significant downstream impacts --- in our case, an inability to create new users or update existing users, which sometimes takes days or longer to determine the root cause and address.

gherceg commented:
Thanks for bringing this to our attention @raytula, and I'm sorry for the inconvenience this caused. In hindsight, there were alternative approaches to rolling this change out that would have minimized/eliminated impact downstream like deprecating email formatted usernames while still allowing successful requests for a period of time, or releasing this change as part of a new API version.

I'm going to add support for email formatted usernames back to this API version as it was a mistake to remove it in the first place.

raytula commented:
Thanks @gherceg 
FYI.  We applied the change on our side and everything seems to be working again. So, no need to change anything further for our specific use case; however, if you think others may be impacted as well, it may be good to support formatted usernames again.

gherceg commented:
> if you think others may be impacted as well, it may be good to support formatted usernames again

Yeah I'm going forward with the understanding that it was a mistake to remove support for formatted usernames, so it is best to add it back. Thanks again for flagging this.

gherceg commented:
FYI @raytula https://github.com/dimagi/commcare-hq/pull/31820 which supports email formatted usernames has been merged into master. Wanted to flag just in case it impacts the patch you have applied on your end.

raytula commented:
> FYI @raytula #31820 which supports email formatted usernames has been merged into master. Wanted to flag just in case it impacts the patch you have applied on your end.

Thanks @gherceg
